File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/performSocialMediaAction/SocialMediaActionPerformHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.performSocialMediaAction

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.performSocialMediaAction.SocialMediaActionPerformApi
import features.performSocialMediaAction.models.PerformSocialMediaActionResponseModel

import features.performSocialMediaAction.serializeSocialMediaActionPerformState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import features.performSocialMediaAction.models.PerformSocialMediaActionRequestModel


class SocialMediaActionPerformHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/perform_social_media_action"

    override val eventChannel: String
        get() = "customer_sdk/events/perform_social_media_action"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return performSocialMediaAction(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun performSocialMediaAction(call: MethodCall): String? {
            val result = SocialMediaActionPerformApi().performSocialMediaAction(
                performSocialMediaActionRequest = Json.decodeFromString(call.argument<String>("performSocialMediaActionRequest") as String),
                businessId = call.argument<String>("businessId") as String
            ).toJsonResult<PerformSocialMediaActionResponseModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        SocialMediaActionPerformApi().socialMediaActionPerformState.serializeSocialMediaActionPerformState().observe { result ->
            pushSerializedResultToEventSink<SocialMediaActionPerformApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/performSocialMediaAction/SocialMediaActionPerformHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class SocialMediaActionPerformHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/perform_social_media_action"

    var eventChannel: String = "customer_sdk/events/perform_social_media_action"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await SocialMediaActionPerformHandler.performSocialMediaAction(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func performSocialMediaAction(call: FlutterMethodCall) async -> String? {
        

        


        
        guard let args = call.arguments else {
            return nil
        }

        do {
           if let myArgs = args as? [String: Any] {
                        guard let performSocialMediaActionRequest = call.arguments as? [String : Any] else {return nil}
                guard let businessId = myArgs["businessId"] as? String else {return nil}

            let result = try await SocialMediaActionPerformApi().performSocialMediaAction(
                methods: performSocialMediaActionRequest,
                businessId: businessId
            ).serializeSocialMediaActionPerformApiResult()

        // bruh
        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
            }
        }
        } catch {
            return nil
        }
        


        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        SocialMediaActionPerformApi().socialMediaActionPerformState
            .serializeSocialMediaActionPerformState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
